fabrics.c

```diff
@@ -505,63 +508,77 @@ int nvmf_discover(const char *desc, int argc, char **argv, bool connect)
-					"command-line options do not match\n",
-					device);
-				c = NULL;
-				persistent = false;
-			} else {
-				/*
-				 * If the controller device is found it must
-				 * be persistent, and shouldn't be disconnected
-				 * on exit.
-				 */
-				persistent = true;
-			}
-		} else {
-			/*
-			 * No controller found, fall back to create one.
-			 * But that controller cannot be persistent.
-			 */
-			fprintf(stderr,
-				"ctrl device %s not found%s\n", device,
-				persistent ? ", ignoring --persistent" : "");
-			persistent = false;
-		}
+	/*
+	 * If no device has been provided or the user insists in a new
+	 * one create a new ctrl.
+	 */
+	if (!device || (persistent && force))
+		goto create_ctrl;
+
+	/* Use provided device as ctrl */
+	c = nvme_scan_ctrl(r, device);
+	if (!c) {
+		ret = errno;
+		goto out_free;
+	}
+
+	/* Check if device matches command-line options */
+	if (!nvme_ctrl_is_discovery_ctrl(c) ||
+	    strcmp(nvme_ctrl_get_transport(c), transport) ||
+	    strcasecmp(nvme_ctrl_get_traddr(c), traddr) ||
+	    (cfg.host_traddr && nvme_ctrl_get_host_traddr(c) &&
+	     strcmp(nvme_ctrl_get_host_traddr(c),
+		    cfg.host_traddr)) ||
+	    (cfg.host_iface && nvme_ctrl_get_host_iface(c) &&
+	     strcmp(nvme_ctrl_get_host_iface(c),
+		    cfg.host_iface)) ||
+	    (trsvcid && nvme_ctrl_get_trsvcid(c) &&
+	     strcmp(nvme_ctrl_get_trsvcid(c), trsvcid))) {
+		fprintf(stderr,
+			"ignoring ctrl device %s, "
+			"command-line options do not match\n",
+			device);
+		nvme_free_ctrl(c);
+	} else {
+		device_found = true;
+		goto discover;
 	}
+
+create_ctrl:
+	/* No device or non-matching device, create a new controller */
+	c = nvme_create_ctrl(r, subsysnqn, transport, traddr,
+			     cfg.host_traddr, cfg.host_iface, trsvcid);
 	if (!c) {
-		/* No device or non-matching device, create a new controller */
-		c = nvme_create_ctrl(r, subsysnqn, transport, traddr,
-				     cfg.host_traddr, cfg.host_iface, trsvcid);
-		if (!c) {
-			ret = errno;
-			goto out_free;
-		}
-		nvme_ctrl_set_discovery_ctrl(c, true);
-		ret = nvmf_add_ctrl(h, c, &cfg);
-		if (ret) {
-			fprintf(stderr,
-				"failed to add controller, error %s\n",
-				nvme_strerror(errno));
-			ret = errno;
-			goto out_free_ctrl;
-		}
+		ret = errno;
+		goto out_free;
+	}
+	nvme_ctrl_set_discovery_ctrl(c, true);
+	ret = nvmf_add_ctrl(h, c, &cfg);
+	if (ret) {
+		fprintf(stderr,
+			"failed to add controller, error %s\n",
+			nvme_strerror(errno));
+		ret = errno;
+		goto out_free_ctrl;
+	}
+
+discover:
+	if (device_found || (persistent && force)) {
+		/*
+		 * If the controller device is found it must be
+		 * persistent, and shouldn't be disconnected on exit.
+		 *
+		 * Also keep the ctrl around if the user insinsts on
+		 * it.
+		 */
+		persistent = true;
+	} else if (persistent) {
+		/*
+		 * No controller found, fall back to create one.
+		 * But that controller cannot be persistent.
+		 */
+		persistent = false;
+		fprintf(stderr,	"ignoring --persistent");
 	}
```

Comment:
Similar here; if we were just setting 'device = NULL' when --force is given we would fallback to the 'normal' code path and create a new device.